### PR TITLE
gogitdiff: incremental build tool

### PR DESCRIPTION
**Pull request by prateek**
- Add a new tool, `ggd`, to compute packages affected by git changes

NB: don't worry about test coverage, this tool uses e2e integration tests using bash to create a separate gopath/git repos (see `build-tools/utilities/ggd/test.sh` for more details).

### CODE REVIEW

File: utilities/ggd/lib/dag.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	}
+
+	graph := make(ImportGraph)
+	for _, pkgInfo := range prog.InitialPackages() {
+		pkg := pkgInfo.Pkg
+		name, imports := pkg.Path(), pkg.Imports()
+		for _, imp := range imports {
+			impPath := imp.Path()
+			if strings.HasSuffix(impPath, "_test") {
+				println(fmt.Sprintf(">> %+v", imp))
```
Comment (prateek) [inline utilities/ggd/lib/dag.go]: dafuq

**Comment by prateek** [comment]
@jeromefroe ended up using your dag calculation code from https://github.com/m3db/build-tools/pull/17, it had minor issues but handled the `x_test` better than https://github.com/kisielk/gotool.

File: utilities/ggd/main.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+		debug("checked out base-branch: %v", baseRef)
+		debug("git-output: %v", checkoutResult)
+		debug("calculating affected packages on base-branch due to changes")
+
+		_, _, baseFullChangeSet, err = dagHelper(packageChanges, *basePkg)
+		dieIfErr(err, "unable to compute affected packages: %v", err)
+		debug("affected packages (including transitive changes): %v", baseFullChangeSet)
+	}
+
+	currentSha1, err := sha1Fn()
```
Comment (jeromefroe) [inline utilities/ggd/main.go]: Should we try to checkout the current SHA again before we return?

also, super nit: `currentSha1` -> `currentSHA1`

File: utilities/ggd/lib/dag.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		g.walk(p, closure)
+	}
+	return closure
+}
+
+func (g ImportGraph) walk(node string, visited ImportSet) {
+	if _, ok := visited[node]; ok {
+		return
+	}
+	visited[node] = struct{}{}
```
Comment (jeromefroe) [inline utilities/ggd/lib/dag.go]: I don't know if we can unconditionally add the package. For example, if we delete a package in our branch then it will be a changed package even though it no longer exists and `go test` will complain that it cant find it.

File: utilities/ggd/README.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
+ggd - GoGitDiff
+===============
+
+This is a tool to compute the packages affected by the difference between two Git
+revisions. It's primary goal is to help speed up CI jobs by figuring out the packages
```
Comment (richardartoul) [inline utilities/ggd/README.md]: "It was deSigned to speed up C.I jobs by determining which packages depend on packages that were changed in the target branch vs. the base branch" or something like that.

Also would be cool to link to Digital Ocean's blog post on the topic

File: utilities/ggd/README.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+affected by git changes, and transitively looking through the Go Imports. It takes
+inspiration from `gta`, and the work done at [DigitalOcean].
+
+NB: debug mode (`-d`) allows users to inspect how the tool arrives at the decisions it does.
```
Comment (richardartoul) [inline utilities/ggd/README.md]: "allows users to inspect the tool's decision making process"

File: utilities/ggd/README.md (hunk 4 above)
Comment (richardartoul) [reply utilities/ggd/README.md]: Edit: nevermind you did link to it

File: utilities/ggd/lib/dag.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+import (
+	"bytes"
+	"fmt"
+	"go/build"
+
+	"golang.org/x/tools/go/buildutil"
+)
+
+// ImportGraph is a map from a package -> all the packages that import it.
```
Comment (richardartoul) [inline utilities/ggd/lib/dag.go]: nit: -> a set of all the packages that import (depend?) on it

File: utilities/ggd/lib/dag.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+type ImportGraph map[string]map[string]struct{}
```
Comment (richardartoul) [inline utilities/ggd/lib/dag.go]: can you use the `ImportSet` as the value for this map?

File: utilities/ggd/lib/dag.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+		}
+	}
+
+	return g, nil
+}
+
+func (g ImportGraph) maybeAddEdge(
+	ctx *build.Context, buildPkg *build.Package, visited map[string]string, pkg, path string,
+) {
+	if path == "C" {
```
Comment (richardartoul) [inline utilities/ggd/lib/dag.go]: I'm confused, what is this?

File: utilities/ggd/README.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
+ggd - GoGitDiff
+===============
+
+This is a tool to compute the packages affected by the difference between two Git
```
[the remark this answers is not in the document]
Comment (prateek) [reply utilities/ggd/README.md]: How about:

```
`ggd` is a tool to compute Go packages affected by changes between two Git revisions. It  does so by identifying the Go packages changed between two Git revisions, and computing all other packages which depend upon them.
```

File: utilities/ggd/lib/dag.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+				// example, when the only go files in a package contain build flags which are
+				// not present in our build context.
+				continue
+			}
+			return nil, fmt.Errorf("could not import package %v: %v", pkg, err)
+		}
+
+		visited := make(map[string]string)
+
+		// ensure we add the pkg to the graph
```
[the remark this answers is not in the document]
Comment (prateek) [reply utilities/ggd/lib/dag.go]: 👍

[the remark this answers is not in the document]
Comment (prateek) [reply utilities/ggd/lib/dag.go]: Yay for tests :)

File: utilities/ggd/lib/dag.go (hunk 6 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: reworded.

File: utilities/ggd/lib/dag.go (hunk 7 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: sure thing

File: utilities/ggd/lib/dag.go (hunk 8 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: CGO .. For context, similar code in go build: https://github.com/golang/go/blob/ae2a2d12f6d8cde35637a13f384f6de524112768/src/go/build/build.go#L860

File: utilities/ggd/README.md (hunk 5 above)
Comment (prateek) [reply utilities/ggd/README.md]: Fair.

File: utilities/ggd/lib/dag.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+func (g ImportGraph) addEdge(from, to string) {
+	edges, ok := g[from]
+	if !ok {
+		edges = make(map[string]struct{})
+		g[from] = edges
+	}
+	edges[to] = struct{}{}
+}
+
+// Closure returns all the transitive closure of all packages reachable
```
[the remark this answers is not in the document]
Comment (prateek) [reply utilities/ggd/lib/dag.go]: fixed

File: utilities/ggd/lib/dag.go (hunk 3 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: Added `testcase7` for the case you're describing, here's the mini-diff: https://github.com/m3db/build-tools/pull/21/commits/f5b1fa0f81ac3c6d44ec1dc2c927bf226dad374b

File: utilities/ggd/main.go [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+	gitChanges, err := lib.ChangedFiles(commitRange, *basePkg)
+	dieIfErr(err, "unable to find changed files from git: %v", err)
+
+	debug("files changed from git: %v", gitChanges)
+	debug("calculating changed packages")
+	packageChanges := lib.ChangedPackages(gitChanges)
+	debug("changed packages: %v", packageChanges)
+
+	baseFullChangeSet := make(lib.ImportSet)
+	if !*includePrechanges {
```
[the remark this answers is not in the document]
Comment (prateek) [reply utilities/ggd/main.go]: The original thought was we'd need to consider both DAGs to answer the question of all affected changes but I'm not convinced that's right anymore. I need to draw some graphs to convince myself before I would be willing to delete the code. Will put down a TODO for now,

File: utilities/ggd/lib/go.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// filterChanges filters changed files based on the following heuristic,
+// only files meeting any of the following conditions affect go compilation
+// 	- any .go file
+//  - any directory called `testdata/`
+//
+// also filters any `vendor/` directory paths
+func filterChanges(input string) bool {
+	if strings.Contains(input, "/vendor/") {
```
Comment (jeromefroe) [inline utilities/ggd/lib/go.go]: Why filter the vendor? Can't we use the DAG to determine exactly what packages depend on them if we strip the vendor prefix?

File: utilities/ggd/lib/dag.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	edges[to] = struct{}{}
+}
+
+// Closure return the transitive closure of all packages reachable
+// by starting at the provided paths in the ImportGraph.
+func (g ImportGraph) Closure(paths ...string) (ImportSet, error) {
+	// add all the paths we're starting at, as they are by definition
+	// reachable.
+	for _, p := range paths {
```
Comment (jeromefroe) [inline utilities/ggd/lib/dag.go]: Instead of including the logic to determine if a package was deleted so we can add all packages here would it make sense to instead only add the package if it exists in the graph since otherwise we know it was deleted?

File: utilities/ggd/lib/go.go (hunk 13 above)
Comment (prateek) [reply utilities/ggd/lib/go.go]: We can, will make it optional. 

I'm thinking of the case for the repos where we include vendor (internally) v the ones we don't (OSS/GH). To your point, it'll be helpful for the former, don't want to include it in the computation for the latter.

File: utilities/ggd/lib/dag.go (hunk 14 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: > "only add the package if it exists in the graph since otherwise we know it was deleted" 

I actually tried to do this the first. Unfortunately, it doesn't work in certain cases. 

Can repro yourself, steps:

```
$ cd $GOPATH/src/github.com/m3db/build-tools
$ git checkout prateek/ggd/demo-testdata-issue
$ cd ./utilities/ggd
$ ./test.sh
# testcase6 will fail
# can see output for the run in ./utilities/ggd/bin/testcase6.out
```

File: utilities/ggd/lib/go.go (hunk 13 above)
Comment (prateek) [reply utilities/ggd/lib/go.go]: done

File: utilities/ggd/lib/go.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// filterChanges filters changed files based on the following heuristic,
+// only files meeting any of the following conditions affect go compilation
+// 	- any .go file
+//  - any directory called `testdata/`
+//
+// also filters any `vendor/` directory paths
+func filterChanges(input string, filterDirPatterns []string) bool {
+	for _, ptrn := range filterDirPatterns {
```
Comment (jeromefroe) [inline utilities/ggd/lib/go.go]: Won't this mean we won't run tests if we just change vendor?

Comment (prateek) [reply utilities/ggd/lib/go.go]: It's an option passed in from the CLI, so up to the user's discretion. 

We'd use it in the following modes:
(a) CI in our internal repo which has `vendor/` checked in, we would not filter `vendor`
(b) CI in our OSS repos (where we do not checkin vendor/`), we would filter `vendor/`

Local development for both would mirror the choices made in the CI.

Comment (jeromefroe) [reply utilities/ggd/lib/go.go]: Wouldn't we then be relying on local CI to verify changes to vendor? I'm kinda of the opinion that if you don't vendor you're dependencies in your repo you need to run tests in CI each time when you pull them.

File: utilities/ggd/lib/dag.go (hunk 14 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: talked offline: if we detect changes in `*/testdata/*`, run everything; otherwise - do ^^ some version of what Jerome's suggesting.

File: utilities/ggd/lib/go.go (hunk 15 above)
Comment (prateek) [reply utilities/ggd/lib/go.go]: done.

File: utilities/ggd/lib/dag.go (hunk 14 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: Done. There's one more edge case we didn't talk about: a dangling package delete. i.e. a package that's not imported by any other packages, and gets deleted. It isn't present the import graph either. I throw an error for it and run tests on all non-filtered packages.

File: utilities/ggd/lib/dag.go [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+	return closure, nil
+}
+
+func (g ImportGraph) walk(node string, visited ImportSet) error {
+	if _, ok := visited[node]; ok {
+		return nil
+	}
+	if _, ok := g[node]; !ok {
+		// NB(prateek): this happens in the case of un-used deleted packages. Look at testcase7.
+		return DanglingDeleteError{node}
```
Comment (jeromefroe) [inline utilities/ggd/lib/dag.go]: Why return an error here?

File: utilities/ggd/main.go [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+	catchallPatterns = getopt.ListLong("catchall-patterns", 'c', "comma separated file/dir names, such that if any of these are changed, all tests are run")
+	filterPatterns   = getopt.ListLong("filter-patterns", 'f', "comma separated dir names to be filtered out during change calculation")
+	dotOutputFile    = getopt.StringLong("output-file", 'o', "", "optionally, visualize changes in a dot file (only in debug mode)")
+)
+
+var (
+	defaultCatchallPatterns = []string{
+		"/testdata/", // go list ./... is unable to gauge impact of changes to this directory
+	}
+	defaultFilterPatterns = []string{"/vendor/"}
```
Comment (jeromefroe) [inline utilities/ggd/main.go]: Should the default for this be empty?

File: utilities/ggd/lib/misc.go [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"path/filepath"
+	"strings"
+)
+
+// ChangedPackages returns all changed packages.
+func ChangedPackages(changedFiles []string) []string {
+	changedPkgsMap := make(map[string]struct{}, len(changedFiles))
+	for _, f := range changedFiles {
+		dir := filepath.Dir(f)
```
Comment (jeromefroe) [inline utilities/ggd/lib/misc.go]: Is it ok to take the whole directory path? Was thinking we might need to potentially remove `GOPATH` if it's in the prefix so we can ensure we have the correct import path.

File: utilities/ggd/lib/dag.go (hunk 16 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: need to indicate to main this happened so we can act on it

File: utilities/ggd/main.go (hunk 17 above)
Comment (prateek) [reply utilities/ggd/main.go]: No. Going to stick with this as the default because it's typically what people want when using the tool by hand. I prefer to optimise for that over what the CI invocation for this will be. 

`go test` does this too so it's not unexpected behaviour.

File: utilities/ggd/lib/misc.go (hunk 18 above)
Comment (prateek) [reply utilities/ggd/lib/misc.go]: the changed files are relative paths already. Will add a comment to indicate.

File: utilities/ggd/lib/dag.go (hunk 16 above)
Comment (jeromefroe) [reply utilities/ggd/lib/dag.go]: Yea, I guess my question is why do we need to act on it? At least, it doesn't seem like an error to me if I remove a package that wasn't imported.

File: utilities/ggd/main.go (hunk 17 above)
Comment (jeromefroe) [reply utilities/ggd/main.go]: Hmm, but if working locally and I update a vendored dependency wouldn't it be preferable to test the packages which depend on the changed vendor package?

File: utilities/ggd/lib/dag.go (hunk 16 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: It's not, but I'm being paranoid and rerun the entire test suite if this situation occurs.

File: utilities/ggd/lib/misc.go (hunk 18 above)
Comment (jeromefroe) [reply utilities/ggd/lib/misc.go]: Hmm, so then the packages in the import statements are relative as well? For example, if I do a git diff in statsdex the paths of files which have changed will be relative to the base of the repo (e.g. `services/collector/etc`), but if I were to import one of those packages they would use the full Git path (e.g. `<hostname>/infra/statsdex/services/collector/etc`).

File: utilities/ggd/lib/dag.go (hunk 16 above)
Comment (jeromefroe) [reply utilities/ggd/lib/dag.go]: I see, maybe worth just calling that out?

File: utilities/ggd/main.go (hunk 17 above)
Comment (prateek) [reply utilities/ggd/main.go]: Sure. But in that case, you've updated `glide.yaml/glide.lock` (added to catchall patterns) and we'd run all tests

File: utilities/ggd/lib/dag.go (hunk 16 above)
Comment (prateek) [reply utilities/ggd/lib/dag.go]: sure

Comment (prateek) [reply utilities/ggd/lib/dag.go]: done

File: utilities/ggd/main.go (hunk 17 above)
Comment (prateek) [reply utilities/ggd/main.go]: or, they can alternatively run with `ggd -f=''` and get the same behaviour (w/o using glide changes as catch all)

File: utilities/ggd/lib/misc.go (hunk 18 above)
Comment (prateek) [reply utilities/ggd/lib/misc.go]: > if I were to import one of those packages they would use the full Git path (e.g. <hostname>/infra/statsdex/services/collector/etc

I don't follow. Go imports are always `GOPATH` or `repo/vendor/` relative, what situation are they not going to be?

**Comment by prateek** [comment]
Boo yah!
